liquid-derive/src/filter_parameters.rs

```diff
@@ -0,0 +1,780 @@
+                }
+                _ => return Err(Error::new_spanned(ty, Self::ERROR_INVALID_TYPE)),
+            },
+            "Expression" => {
+                if !path.arguments.is_empty() {
+                    return Err(Error::new_spanned(ty, Self::ERROR_INVALID_TYPE));
+                } else {
+                    return Ok(false);
+                }
+            }
+            _ => return Err(Error::new_spanned(ty, Self::ERROR_INVALID_TYPE)),
+        }
+    }
+
+    /// Creates a new `FilterParameter` from the given `field`, with the given `name`.
+    fn new(name: &'a Ident, field: &Field) -> Result<Self> {
+        let is_optional = Self::parse_type_is_optional(&field.ty)?;
+        let meta = FilterParameterMeta::from_field(&field)?;
+
+        Ok(FilterParameter {
+            name,
+            is_optional,
+            meta,
+        })
+    }
+
+    /// Returns whether this field is optional.
+    fn is_optional(&self) -> bool {
+        self.is_optional
+    }
+
+    /// Returns whether this field is required (not optional).
+    fn is_required(&self) -> bool {
+        !self.is_optional
+    }
+
+    /// Returns whether this is a positional field.
+    fn is_positional(&self) -> bool {
+        self.meta.mode == FilterParameterMode::Positional
+    }
+
+    /// Returns whether this is a keyword field.
+    fn is_keyword(&self) -> bool {
+        self.meta.mode == FilterParameterMode::Keyword
+    }
+
+    /// Returns the name of this parameter in liquid.
+    ///
+    /// That is, by default, the name of the field as a string. However,
+    /// this name may be overriden by `rename` attribute.
+    fn liquid_name(&self) -> String {
+        match &self.meta.rename {
+            Some(name) => name.clone(),
+            None => self.name.to_string(),
+        }
+    }
+}
+
+impl<'a> ToTokens for FilterParameter<'a> {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        self.name.to_tokens(tokens);
+    }
+}
+
+/// Whether `FilterParameter` is `Keyword` or `Positional`.
+#[derive(PartialEq)]
+enum FilterParameterMode {
+    Keyword,
+    Positional,
+}
+
+impl FromStr for FilterParameterMode {
+    type Err = String;
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            "keyword" => Ok(FilterParameterMode::Keyword),
+            "positional" => Ok(FilterParameterMode::Positional),
+            s => Err(format!(
+                "Expected either \"keyword\" or \"positional\". Found \"{}\".",
+                s
+            )),
+        }
+    }
+}
+
+/// The type that this `FilterParameter` will evaluate to.
+enum FilterParameterType {
+    // Any value, the default type
+    Value,
+
+    // Scalars
+    Integer,
+    Float,
+    Bool,
+    Date,
+    Str,
+}
+
+impl FromStr for FilterParameterType {
+    type Err = String;
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            "any" => Ok(FilterParameterType::Value),
```

Comment:
imo the macro parameter `value="human description"` should either be `type="rust type"` or `type="liquid value type"`.  Things like "whole number" are to be less jargon-y for the end-user while here our target is a developer and we should use more concrete, discoverable names.

Reply:
Yeah I'm not a big fan of those terms either. I didn't really know what I would do here. On one side, using rust types could be confusing because people could erroneously thing that any rust type could be put there. The liquid types I didn't really know what were the 'official' terms for each type. Guess I'll use the names of scalar enum: `any, integer, float, bool, date, str`.

As for `value` vs `type`, I initially also started with `type`. However, because type is a keyword in rust, for some reason it doesn't let me use it inside the macro. I'm not sure if that is a limitation of macros themselves or in the implementation of the `syn` crate.

```
error: expected identifier, found keyword `type`
  --> src\filters\mod.rs:73:59
   |
73 |     #[parameter(description = "The offset of the slice.", type = "whole number")]
   |                                                           ^^^^ expected identifier, found keyword
```

Reply:
If nothing else, could just be that they wanted the parser to be as context-free as possible.

Yeah, I was thinking scalar enum variants.  Some of the filters take Array and Object a well.  Not sure if you were also planning on value variants or just having thos be "value".

Hmmm, not sure what to do about `type`.  `value` throws me off because I expect a value with it.  

Alternatively, we could switch the derive to be on the `Arguments` and to use the data type the user specifies.  I don't remember why I had it on the `Parameters`.  I think its because I generally hate idents that are generated by macros and so that was the one I wanted more visible?

Reply:
I decided not to add Array or Object because those are more complex cases. I see this as an extra shortcut to reduce boilerplate in checking the type of the arguments (in a way, almost like to compensate for the extra boilerplate of the new API in general). What I mean is that people can perfectly opt out of using this element (which defaults to "any") and do the check on their own. For complex cases (for example, an argument that can be either a number or a string; an argument that can be an Object) I think there would be too many cases to cover, so the user is responsible to implement them manually. In the future, however, if one of those cases becomes too common, a new shortcut can always be added for it.

Regarding type vs value, I've just had the sudden idea to call it something like arg_type. It would be a little longer but at least it's more clear what it means.

As for switching the derive to be on the `EvaluatedFilterParameters`, I'm not a fan of that idea. The user would then have to refer to the name of a structure that doesn't exist within their code, which would feel confusing in my opinion.

Reply:
> In the future, however, if one of those cases becomes too common, a new shortcut can always be added for it.

Alternatively, we can take the `structopt` approach and accept a function name to call to process the data.

> As for switching the derive to be on the `EvaluatedFilterParameters`, I'm not a fan of that idea. The user would then have to refer to the name of a structure that doesn't exist within their code, which would feel confusing in my opinion.

Though there is a similar degree of weirdness that the user is accessing fields in a type that they can't see.  This is why I somewhat lean towards rust types so its very explicit and less guess work is involved.

Reply:
>Though there is a similar degree of weirdness that the user is accessing fields in a type that they can't see. This is why I somewhat lean towards rust types so its very explicit and less guess work is involved.

On the other hand, this would make it feel like it would accept any rust type, be it i32, i64, i16, usize, etc..., while the intention is in a way to "unwrap" the `Value` struct into de corresponding scalar, returning whatever that scalar has as a type.